keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -847,69 +847,156 @@ Because we now have more possible values for `mu(i,t)` than 0 and
+from the variations in `rate`.  This is similar to the use of `R`
+values in the original Fair Queuing scheme.
+
+```
+PEnd(i,j) = P(i, B(i,j)) + len(i,j)
+```
+
+For a given queue `i` at a given time `t` we can write the expected
+end (EE) time of each active packet `j` as the current time plus the
+expected amount of time needed to transmit the bits that have not
+already been transmitted (making the assumption that the current rate
+will continue into the future):
+
+```
+EE(i,j,t) = t + (PEnd(i,j) - P(i,t)) / rate(i,t)
+```
+
+Notice that the remaining time to transmit the packet, `EE(i,j,t)-t`,
+is a function of:
+- a packet-specific quantity (`PEnd`) that does not change over time,
+  and
+- queue-specific quantities (`P`, `rate`) that change over time and
+  are independent of packet.
+
+The complexity of updating this representation of a queue's expected
+end times to account for the passage of time or a change in `mu` does
+not require modifying the packet-specific data (`PEnd` values) in this
+data structure, thus costs `O(1)`.  Adding or removing a packet or
+changing its length (see below) does not require changing the
+packet-specific data of the other active packets.
+
+We can keep the active packets of a queue in a logarithmic-complexity
+sorted data structure ordered by expected end time.  Adding or
+removing a packet from the active set or changing the packet's length
+will cost O(log(size of the active set)).
+
+We can divide the non-empty queues into two categories and keep each
+in its own data structure.  For the queues that get `mu(i,t) ==
+rho(i,t)`, keep them in a logarithmic-complexity sorted data structure
+ordered by the earliest expected end time of the queue's active
+packets.  Changes to `mu_fair` do not affect this data structure,
+except to the degree that they cause queues to enter or leave this
+category.
+
+Similarly, we can keep the queues for which `mu(i,t) == mu_fair(t)` in
+another sorted data structure ordered by earliest expected end time.
+Since `mu(i,t)` is the same for all queues in this category, the
+passage of time and changes in `mu_fair` do not change the ordering of
+packets or queues in this data structure, except to the degree that
+queues enter or leave this category.  The representation of expected
+end times in this category gets one more level of indirection, through
+that shared `mu_fair`.
+
+When a change in `mu_fair` causes `n_delta` queues to move from one
+category to another, it costs `O(n_delta * log num_queues)` to update
+these data structures by those moves.
+
+Updating the data structures for a mere change in one queue's `NAP`
+has logarithmic cost.
+
+The above discussion concerns the virtual world, which transmits each
+packet no more quickly than the real world.  Usually a packet will
+finish transmission in the real world before it finishes in the
+virtual word.  But it is important to keep each packet in the virtual
```

Comment:
What if a packet finishes earlier in the virtual world? Do we need to adjust something?

Reply:
You mean what if a packet finishes earlier in the virtual world than in the real world?  That possibility was explicitly denied by the first sentence in this paragraph.  The reason is that `rate(i,t) <= mu_single`.

Reply:
Let me rephrase my question. I understand that in this whole proof it works.

I was more asking for "practical consequences". Let's say that I incorrectly configured mu_single (inflight requests). Or because some hardware or whatever reasons, my request actually took more than it should.

Are we just ignoring that? Or do we need to take any actions in this case?

Reply:
In the worl of serving requests, `mu_single` is always exactly one seat --- not something that can be misconfigured.  What is questionable is `C` --- the number of available seats.  We have long talked about auto-tuning `C`.  That's a future issue.  We also talked on Monday about recognizing that these servers do not really serve from one homogeneous underlying resource, but had no good ideas about a practical way to cope with this.  However, I will add that I think the special treatment for longrunning requests is, at least in part, a coarse way of dealing with that.

Reply:
Some thoughts before I forget... (only slightly related to this comment thread)

* it occurs to me that if we did have a width value for requests (that is, the number of seats the request occupies), it doesn't need to be an integer. E.g. it's quite plausible that 2 big requests collectively occupy 3 seats-- 1.5 each.
* Walking back what I said in the meeting-- I don't think width and length trade off 1:1 for all requests.
* However I think width and length trade off at least a little bit (e.g., if you serve fewer requests simultaneously, we'd expect a tiny bit less latency for the requests you are serving, under a wide variety of assumptions about the server)
* We can always pretend, if we wish, that a request takes longer to serve than it actually does in reality.
* So, perhaps we can be willing to artificially narrow a request (reduce it's width) by a factor of X in exchange for increasing its length by a factor of X*Y, where Y > 1; and we guarantee that the request will actually occupy its (narrower) seat for that long.

I think the net effect of this line of thinking implies that
* we don't need an integer number of seats (!)
* situations where the seats are narrower on average than the requests degrade into a duty cycle
* To the algorithm you are describing in this PR, the only modification is that instead of looping while there is an empty seat, we loop until the next request to serve is wider than the remaining amount of seat. (maybe we should call it a bench at this point)
* If there are zero occupants of the bench, and the next packet is too wide, then we can reduce its width and lengthen it. The artificial length prevents us from doing this too often. Perhaps the zero occupant constraint can be relaxed. We need to take care not to do something like make a request 1/10th as wide in exchange for making it take a thousand years.

Reply:
> The current revision of this PR talks about revising the guessed service duration for requests as they prove they need more time.


Thanks @MikeSpreitzer - this looks fine.

[Also to explain my previous (not well articulated thoughts) - I wasn't really questioning the math (although I really like the current version which makes it very clear why it holds). I was more questioning the "perfectness of the worlds".
In other words - even though if in 99,99% of cases our guesses are fine, it can happen that e.g. something will stuck in the network between kube-apiserver and etcd and it will actually take more time than it should.

The current explicitly said " the guess is increased" is what fully satisfies me now.


Re @lavalamp `s thought - I started thinking about non-integer number of seats myself somewhat independenty and came to the same conclusion that we should be possible.
That said, I think this PR is already provides a strict improvement, so I'm not sure we should block this PR on accomodating those thoughts.
@MikeSpreitzer - how about just adding Daniel`s thoughts in somewhat "to-think-through in more details" paragraph?

[

Reply:
I think we should incorporate my thoughts in a follow-up PR, not make this one bigger :)

Reply:
> we loop until the next request to serve is wider than the remaining amount of seat

BTW I thought about this more and if the requests are different widths, then one could imagine wanting a search routine to find a good fit rather than smooshing the soonest-finishing request (which would probably make it not the soonest-finishing request any more, anyway!)-- like maybe you want to search for a request that's currently scheduled to finish at the resulting lengthed time anyway and start working on it now.

I think this can be viewed as an optimization problem where we desire a heuristic that e.g.
* can be evaluated in O(log N) time
* Averages out to max-min fairness (i.e., the mistakes are not always biased in favor of some type of queue or request) over relatively short timeframes

But that invites the question of what exactly we're optimizing :) I think implicitly the optimization target in this PR is "keep the server 100% busy if there's any work to be done". We can also imagine other targets, such as "minimize 99%ile latency" or "minimize change in server resource consumption" (i.e., slow down utilization spikes), or some combination of all of these, and I'm sure there's other reasonable targets, too.

The implication of this (if any) might be that we should keep the code modular to the point where it's easy to try out and measure alternative heuristics.

Reply:
I thought it was obvious that width does not have to be an integer.

If we suppose requests have different widths, we have to think harder about what fairness really means: what if a wide request holds up dispatching narrow ones?  Or vice versa?

I think being "work-conserving" is a top priority.  Anything less will just piss people off.  After that comes differentiating traffic into priority levels and giving fairness between flows.

In an idealized world we might be able to argue that since the product of width and length is the measure of work in a request, we only need to consider this product and not its factors.  But I think that ignores an important fact about real servers: their throughput crashes if you ask them to do too much work concurrently.  The scheme we currently have, of regulating concurrency, has the virtue of respecting this real limitation.

In other words, the point of width is to characterize the bite that a request takes out of a server's ability to serve requests concurrently.  Thus far we have been supposing that all requests take an equal size bite (except the longrunning requests, which we suppose take zero bite --- obviously a point-in-time simplification of a more complex situation).  This is a simplification of what we know is a very complex situation: serving a request generally involves loading various subsystems over various intervals of time; in other words, serving a request takes many smaller bites out of many underlying resources.  It seems to me that to justify having a single simple concept of "width" implies the following model: serving a request draws on two resources, one of which can support limited concurrency and the other of which can support as much concurrency as will ever actually be thrown at it (given the limits imposed by the first).  For a very simple example, a concept of width would apply if every request did some mixture of single-CPU work and sleeping; the width would be the fraction of its service time that the request spends using that single CPU.  If we really do think that we have such a bi-modal underlying resource usage pattern, then a concept of width would be a good way to manage it.

I agree with putting further complexities in follow-on PRs.